File: 10. ShellSort.py
def shell_sort(arr):
    n = len(arr)
    
    # Start with a big gap, then reduce the gap
    gap = n // 2
    while gap > 0:
        # Perform a gapped insertion sort for this gap size.
        # The first gap elements arr[0..gap-1] are already in gapped order
        # so we can start the loop from gap.
        for i in range(gap, n):
            # Save arr[i] in temp and make a hole at position i
            temp = arr[i]
            
            # Shift earlier gap-sorted elements up until the correct location
            # for arr[i] is found
            j = i
            while j >= gap and arr[j - gap] > temp:
                arr[j] = arr[j - gap]
                j -= gap
            
            # Put temp (the original arr[i]) in its correct location
            arr[j] = temp
            
        print(f"After gap {gap}: {' '.join(map(str, arr))}")
        
        # Reduce the gap for the next iteration
        gap //= 2

def main():
    """Main function to test the shell_sort."""
    arr = [12, 34, 54, 2, 3]

    print(f"Original array: {' '.join(map(str, arr))}")

    shell_sort(arr)

    print(f"Sorted array: {' '.join(map(str, arr))}")

if __name__ == "__main__":
    main()
    """
    Sorts a list in place using the Shell Sort algorithm.

    Type: Comparison-based, Not stable, In-place

    Time Complexity (Worst Case):
    Depends on the gap sequence used.
    - Original Shell’s sequence (n/2, n/4, ...): O(n^2)
    - Better sequences (e.g., Sedgewick): Can reach O(n^(4/3))
    - Best Case: O(n log n)

    Space Complexity:
    - Auxiliary Space: O(1) -> In-place sorting
    """

File: 10. ShellSort.cpp
#include <bits/stdc++.h>
using namespace std;

void shellSort(vector<int> &arr) {
    int n = arr.size();

    // Start with a big gap, then reduce the gap
    for(int gap = n/2; gap > 0; gap /= 2) {
        // Do a gapped insertion sort for this gap size
        for(int i = gap; i < n; i++) {
            int temp = arr[i];
            int j = i;
            while(j >= gap && arr[j - gap] > temp) {
                arr[j] = arr[j - gap];
                j -= gap;
            }
            arr[j] = temp;
        }
        // Display array after current gap
        cout << "After gap " << gap << ": ";
        for(int num : arr) cout << num << " ";
        cout << endl;
    }
}

int main() {
    vector<int> arr = {12, 34, 54, 2, 3};

    cout << "Original array: ";
    for(int num : arr) cout << num << " ";
    cout << endl;

    shellSort(arr);

    cout << "Sorted array: ";
    for(int num : arr) cout << num << " ";
    cout << endl;

    return 0;
}

/*
Type: Comparison-based, Not stable, In-place

Time Complexity (Worst Case):
Depends on the gap sequence:
Original Shell’s gaps: O(n²)
Using better sequences (e.g., Hibbard, Sedgewick): O(n^(3/2)) to O(n^(4/3))
Best Case: O(n log n) (for some sequences, nearly sorted array)
Average Case: ~O(n^(3/2)) for common sequences

Space Complexity:
Auxiliary Space: O(1) → In-place

*/